src/VisualStudio/Core/Def/Implementation/Utilities/ServiceInitializer.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Utilities
+{
+    internal class ServiceInitializer<TInterface, TService> where TInterface : class
+    {
+        private readonly AsyncLazy<TInterface> _asyncLazy;
+
+        public ServiceInitializer(Shell.IAsyncServiceProvider serviceProvider, Func<TInterface, Task> initializer, bool uiThreadRequired, JoinableTaskFactory joinableTaskFactory)
+        {
+            _asyncLazy = new Threading.AsyncLazy<TInterface>(async () =>
```

Comment:
@AArnott once https://github.com/Microsoft/vs-threading/pull/317 get checked in, I might just use it directly rather than this wrapper.

Reply:
You might also like `AsyncLazyInitializer`, which is also [coming soon](https://github.com/AArnott/vs-threading/commit/b4c07f334e4b53c34340604fef001dcaf0f639d9).